tests/unit/maze_transformer/generation/test_tokenizers.py

```diff
@@ -1,5 +1,276 @@
+from maze_transformer.training.tokenizer import HuggingMazeTokenizer
+from scripts.create_dataset import generate_MazeTokenizer
+
+
+def test_tokenization_encoding():
+    # Check that wrapped tokenizer __call__ returns the same as original tokenizer
+    maze = generate_MazeTokenizer(None, 3, (0, 0), (2, 1))
+
+    # Need to generate a config to extract the token map >.<
+    cfg = MazeDatasetConfig(name="testing_maze", grid_n=3, n_mazes=1)
+    node_token_map = cfg.node_token_map
+
+    # Adjacency List Tokenization
+    maze_str_tokens = maze.as_tokens(node_token_map)
+
+    # Manual Tokenization
+    vocab_map = {k: v for v, k in enumerate(cfg.token_arr)}
+    maze_tokens = [vocab_map[token] for token in maze_str_tokens]
+
+    # WrappedTokenizer
+    # Initialized with a configholder - tokenizer will eventually be a string
+    cfg_holder = ConfigHolder(
+        train_cfg=None, dataset_cfg=cfg, model_cfg=None, tokenizer=None
+    )
+    tokenizer = HuggingMazeTokenizer(cfg_holder)
+
+    tokenizer_out = tokenizer(maze_str_tokens)["input_ids"]
+    assert torch.all(
+        torch.tensor(tokenizer_out).flatten() == torch.tensor(maze_tokens)
+    ), "Tokenization mismatch"
+
+
+def test_ascii_encoding():
+    # Check that the ascii encoding works for multiple different inputs
+    maze_str_tokens = [
+        "<ADJLIST_START>",
+        "(1,1)",
+        "<-->",
+        "(2,1)",
+        ";",
+        "(2,0)",
+        "<-->",
+        "(1,0)",
+        ";",
+        "(0,1)",
+        "<-->",
+        "(0,0)",
+        ";",
+        "(2,2)",
+        "<-->",
+        "(2,1)",
+        ";",
+        "(2,0)",
+        "<-->",
+        "(2,1)",
+        ";",
+        "(0,2)",
+        "<-->",
+        "(1,2)",
+        ";",
+        "(0,0)",
+        "<-->",
+        "(1,0)",
+        ";",
+        "(0,2)",
+        "<-->",
+        "(0,1)",
+        ";",
+        "<ADJLIST_END>",
+        "<TARGET_START>",
+        "(2,1)",
+        "<TARGET_END>",
+        "<START_PATH>",
+        "(0,0)",
+        "(1,0)",
+        "(2,0)",
+        "(2,1)",
+        "<END_PATH>",
+    ]
+    target = [
+        "#######",
+        "#     #",
+        "# ### #",
+        "# #   #",
+        "# ### #",
+        "#   # #",
+        "#######",
+    ]
+
+    # Need to generate a config to extract the token map >.<
+    cfg = MazeDatasetConfig(name="testing_maze", grid_n=5, n_mazes=1)
+    cfg_holder = ConfigHolder(
+        train_cfg=None, dataset_cfg=cfg, model_cfg=None, tokenizer=None
+    )
+    tokenizer = HuggingMazeTokenizer(cfg_holder)
+
+    # Try with string tokens
+    assert (
+        tokenizer.to_ascii(maze_str_tokens).splitlines() == target
+    ), "ASCII encoding from string tokens failed"
+
+    # And with token ids
+    token_ids = tokenizer.encode(maze_str_tokens)
+    assert (
+        tokenizer.to_ascii(token_ids).splitlines() == target
+    ), "ASCII encoding from token ids failed"
+
+
+def test_inside_hooked_transformer():
```

Comment:
Test a single scenario in each unit test:
- allows for more parallelisation = tests run faster
- tests can have more specific names so it's faster to understand what went wrong when they fail (those comments with the -- -- formatting could be test names)
- if single purpose tests are hard to write it's a useful signal that there's something wrong with the code
- lower risk of state-related bugs from other test scenarios, e.g. if you set a variable correctly in the first scenario, then a bug in the second scenario means it doesn't get set when it's supposed to, the test might not catch it